### PR TITLE
Add a search engine into Plume

### LINKED ISSUE #149: Search

### PR DESCRIPTION
**Pull request by fdb-hiroshima**
Add support for search through [tantivy](https://crates.io/crates/tantivy)
fix #149 

This is mostly done, current issues are:
 - ~~query parser might be [a bit intolerant](https://tantivy-search.github.io/tantivy/tantivy/query/struct.QueryParser.html#method.parse_query), we should probably implement our own~~
- ~~closing Plume leave the database blocked, currently you have to run plm search unlock -p <path to plume wokingdir> to unlock it.~~
 - frontend is done by me (which means it's probably not beautiful), and is not translated
 - ~~advanced search is working, but yet undocumented (i.e. filtering by date could be done if you know enough tantivy, but I don't expect common user to)~~
 - As it has grown more than I anticipated, I should probably add some more tests

After upgrading to this, you will have to run `plm search init -p <path to plume wokingdir>` to initialize search database, otherwise Plume will panic at launch

### CODE REVIEW

**Comment by fdb-hiroshima** [comment]
Small todo list of things that are not strictly needed atm, but should still be done:
 - [ ] split `plume-models/src/search.rs` into multiple files, under a new directory : it's "only" 670 lines, but they are very dense, and although all is related to search, having a sort of database connection, a query parser, and a tokenizer in the same file seems too much
 - [ ] comment the code, especially the parser
 - [ ] maybe make some part less generics but more logical. Currently if searching for author `user1@plume1.org` or `user2@plume2.org`, a post from `user2@plume1.org` would match.
 - [ ] add some tests at least for the parser

**Comment by fulmicoton** [comment]
> query parser might be a bit intolerant, we should probably implement our own

@fdb-hiroshima We have a dangling PR to add a lenient mode to search.

> i.e. filtering by date could be done if you know enough tantivy, but I don't expect common user to)

@fdb-hiroshima  Let me know if you have any question  about tantivy.

**Comment by fdb-hiroshima** [comment]
@fulmicoton Both "issues" are fixed (you can see they are ~~crossed out~~), I've implemented a parser which fit with our exact needs (we know more about what each field represent than Tantivy can).

Concerning dates, I don't have any question, it's just how we give date to tantivy is not evident to the end user, but tantivy can't change that. We store date the result of `num_days_from_ce()`, basically the number of days since some point. This is because Tantivy's RangeQuery `iterate over the terms within the range`, so I tried to have the smallest ranges possible, with as few useless/empty elements as possible (iterating over 30 days vs 2,592,000 seconds, one seems faster than the other).

I really like your crate, it's very complete yet relatively easy to use. The only issue I had with it was about BooleanQuery, I had a hard time understanding how to build theme. I was looking for some kind of function, and took a lot of time to figure out it implement From. Maybe it could just simply be mentioned in the documentation of `BooleanQuery::new_multiterms_query`, "to create a BooleanQuery from Box&#x3C;Query&#x3E; use From<Vec<(Occur, Box<Query>)>> instead" or something like that

**Comment by fulmicoton** [comment]
Opened [tantivy-search/tantivy#446](https://github.com/tantivy-search/tantivy/issues/446) to make creation of `BooleanQuery` easier.

Your insight about range is correct. Eventually tantivy will be a bit smarter than what it does right now, but too be honest this is pretty low priority.

Looking forward to seeing Plume grow.

File: templates/search/index.html.tera [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+Search
+{% endblock title %}
+
+{% block head %}
+<script>
+window.onload = function(evt) {
+	var form = document.getElementById('form');
+	form.addEventListener('submit', function () {
+	    for (var input of form.getElementsByTagName('input')) {
+		if (input.name && !input.value) {
```
Comment (BaptisteGelez) [inline templates/search/index.html.tera]: You can add this condition before to fix your issue:

```js
if (input.name === '') {
    input.name = input.id
}
```

Comment (fdb-hiroshima) [reply templates/search/index.html.tera]: I'll do it then

File: plume-models/src/search/query.rs [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+            self.from_str_req(query)
+        }
+    }
+
+    // map a token and it's field to a query
+    fn token_to_query(token: &str, field_name: &str) -> Box<Query> {
+        let token = token.to_lowercase();
+        let token = token.as_str();
+        let field = Searcher::schema().get_field(field_name).unwrap();
+        if token.contains('@') && field_name=="author" || field_name=="blog" {
```
[the remark this answers is not in the document]
Comment (fdb-hiroshima) [reply plume-models/src/search/query.rs]: Usually && has the priority, which mean this probably does not do what I expect. I'll verify if it's working as intended

[the remark this answers is not in the document]
Comment (fdb-hiroshima) [reply plume-models/src/search/query.rs]: I confirm searching a blog with no domain panic at next line

**Comment by fdb-hiroshima** [comment]
@BaptisteGelez in what order do we merge search ructe and the dependency upgrade?

**Comment by fdb-hiroshima** [comment]
I'll do so

**Comment by fdb-hiroshima** [comment]
There is one thing I forgot. There is nowhere a link to `/search`

**Comment by BaptisteGelez** [comment]
You're right, but I'll a a link the header when converting the template to ructe. And updating the docs can be done in another PR.

**Comment by fdb-hiroshima** [comment]
if you want you can make a small form with an `<input name="q">`, it is enough to make a simple query, and would probably fit better with the ergonomic people expect (being able to do a basic search from anywhere, and an advanced search if the result were inconclusive)

**Pull request merged**